File: src/remedy/interfaces/xai_auth.py
"""xAI (Grok) authentication: API key + OAuth device-code (OpenCode-style).

API key path uses console keys (XAI_API_KEY / xai-…).
OAuth path uses accounts.x.ai device authorization for SuperGrok / X Premium+.

OAuth client id defaults to the public Grok Build / open-agent ecosystem client
used by OpenClaw and peers. Override with REMEDY_XAI_OAUTH_CLIENT_ID if you
register your own application with xAI.
"""

from __future__ import annotations

import json
import logging
import os
import threading
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

AUTH_SERVER = "https://accounts.x.ai"
API_BASE = "https://api.x.ai/v1"
DEVICE_CODE_URL = f"{AUTH_SERVER}/oauth2/device/code"
TOKEN_URL = f"{AUTH_SERVER}/oauth2/token"
# Public client id used by Grok Build / OpenClaw-class agents for device OAuth.
# https://github.com/openclaw/openclaw/issues/84504
DEFAULT_CLIENT_ID = "b1a00492-073a-47ea-816f-4c329264a828"
GRANT_DEVICE = "urn:ietf:params:oauth:grant-type:device_code"

_refresh_lock = threading.Lock()
_poll_sessions: dict[str, dict[str, Any]] = {}
_poll_lock = threading.Lock()


def _client_id() -> str:
    return (
        os.environ.get("REMEDY_XAI_OAUTH_CLIENT_ID", "").strip()
        or DEFAULT_CLIENT_ID
    )


def auth_dir(home: Path | None = None) -> Path:
    base = Path(home or os.environ.get("REMEDY_HOME", "~/.remedy")).expanduser()
    d = base / "auth"
    d.mkdir(parents=True, exist_ok=True)
    return d


def auth_path(home: Path | None = None) -> Path:
    return auth_dir(home) / "xai.json"


@dataclass
class XaiCredentials:
    auth_method: str = "none"  # none | api_key | oauth
    api_key: str | None = None
    access_token: str | None = None
    refresh_token: str | None = None
    expires_at: float | None = None  # unix seconds
    token_type: str = "Bearer"
    raw: dict[str, Any] = field(default_factory=dict)

    @property
    def connected(self) -> bool:
        if self.auth_method == "api_key" and self.api_key:
            return True
        if self.auth_method == "oauth" and self.access_token:
            if self.expires_at and time.time() >= float(self.expires_at) - 60:
                return bool(self.refresh_token)
            return True
        return False

    def bearer_token(self) -> str | None:
        """Token suitable for Authorization: Bearer …"""
        if self.auth_method == "api_key" and self.api_key:
            return self.api_key
        if self.auth_method == "oauth" and self.access_token:
            return self.access_token
        return None

    def to_public_dict(self) -> dict[str, Any]:
        return {
            "provider": "xai",
            "auth_method": self.auth_method,
            "connected": self.connected,
            "has_api_key": bool(self.api_key),
            "has_oauth": bool(self.access_token or self.refresh_token),
            "expires_at": self.expires_at,
        }


def load_credentials(home: Path | None = None) -> XaiCredentials:
    path = auth_path(home)
    if not path.exists():
        # Env bootstrap
        env_key = (
            os.environ.get("XAI_API_KEY", "").strip()
            or os.environ.get("REMEDY_XAI_API_KEY", "").strip()
        )
        if env_key:
            return XaiCredentials(auth_method="api_key", api_key=env_key)
        return XaiCredentials()
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return XaiCredentials()
    return XaiCredentials(
        auth_method=str(data.get("auth_method") or "none"),
        api_key=data.get("api_key"),
        access_token=data.get("access_token"),
        refresh_token=data.get("refresh_token"),
        expires_at=data.get("expires_at"),
        token_type=str(data.get("token_type") or "Bearer"),
        raw=data,
    )


def save_credentials(creds: XaiCredentials, home: Path | None = None) -> None:
    path = auth_path(home)
    payload = {
        "auth_method": creds.auth_method,
        "api_key": creds.api_key,
        "access_token": creds.access_token,
        "refresh_token": creds.refresh_token,
        "expires_at": creds.expires_at,
        "token_type": creds.token_type,
        "updated_at": time.time(),
    }
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    try:
        path.chmod(0o600)
    except OSError:
        pass


def clear_credentials(home: Path | None = None) -> None:
    path = auth_path(home)
    if path.exists():
        path.unlink()


def save_api_key(api_key: str, home: Path | None = None) -> XaiCredentials:
    key = (api_key or "").strip()
    if not key:
        raise ValueError("API key is empty")
    creds = XaiCredentials(auth_method="api_key", api_key=key)
    save_credentials(creds, home=home)
    return creds


def _http_form(url: str, data: dict[str, str], timeout: float = 30.0) -> dict[str, Any]:
    body = urllib.parse.urlencode(data).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json",
            "User-Agent": "RemedyDesktop-xAI-Auth/1.0",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read().decode("utf-8")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8", errors="replace")
        try:
            parsed = json.loads(err_body)
        except json.JSONDecodeError:
            parsed = {"error": err_body or str(e), "status": e.code}
        parsed.setdefault("status", e.code)
        raise RuntimeError(json.dumps(parsed)) from e


def start_device_login(home: Path | None = None) -> dict[str, Any]:
    """Begin OAuth device-code flow. Returns user_code + verification URLs."""
    client_id = _client_id()
    data = _http_form(
        DEVICE_CODE_URL,
        {
            "client_id": client_id,
            "scope": "openid profile email offline_access",
        },
    )
    device_code = data.get("device_code")
    user_code = data.get("user_code")
    if not device_code or not user_code:
        raise RuntimeError(f"Unexpected device-code response: {data}")

    interval = int(data.get("interval") or 5)
    expires_in = int(data.get("expires_in") or 900)
    verification_uri = (
        data.get("verification_uri")
        or data.get("verification_url")
        or f"{AUTH_SERVER}/oauth2/device"
    )
    verification_uri_complete = data.get("verification_uri_complete") or (
        f"{verification_uri}?user_code={urllib.parse.quote(str(user_code))}"
    )

    session_id = str(user_code)
    with _poll_lock:
        _poll_sessions[session_id] = {
            "device_code": device_code,
            "client_id": client_id,
            "interval": interval,
            "expires_at": time.time() + expires_in,
            "home": str(home) if home else None,
            "status": "pending",
            "error": None,
        }

    # Background poller
    threading.Thread(
        target=_poll_until_done,
        args=(session_id,),
        daemon=True,
        name="xai-oauth-poll",
    ).start()

    return {
        "session_id": session_id,
        "user_code": user_code,
        "device_code": device_code,
        "verification_uri": verification_uri,
        "verification_uri_complete": verification_uri_complete,
        "interval": interval,
        "expires_in": expires_in,
        "status": "pending",
        "message": (
            f"Open {verification_uri_complete} and approve access, "
            f"or go to {verification_uri} and enter code {user_code}."
        ),
    }


def _poll_until_done(session_id: str) -> None:
    with _poll_lock:
        sess = dict(_poll_sessions.get(session_id) or {})
    if not sess:
        return
    device_code = sess["device_code"]
    client_id = sess["client_id"]
    interval = max(3, int(sess.get("interval") or 5))
    expires_at = float(sess.get("expires_at") or (time.time() + 900))
    home = Path(sess["home"]) if sess.get("home") else None

    while time.time() < expires_at:
        time.sleep(interval)
        try:
            data = _http_form(
                TOKEN_URL,
                {
                    "grant_type": GRANT_DEVICE,
                    "device_code": device_code,
                    "client_id": client_id,
                },
            )
        except RuntimeError as e:
            try:
                err = json.loads(str(e))
            except json.JSONDecodeError:
                err = {"error": str(e)}
            err_code = str(err.get("error") or err.get("error_description") or "")
            if "authorization_pending" in err_code or "slow_down" in err_code:
                if "slow_down" in err_code:
                    interval = min(interval + 2, 15)
                continue
            if "expired" in err_code or "access_denied" in err_code:
                with _poll_lock:
                    if session_id in _poll_sessions:
                        _poll_sessions[session_id]["status"] = "error"
                        _poll_sessions[session_id]["error"] = err_code
                return
            # Unknown error — keep pending briefly
            logger.debug("xAI token poll error: %s", err)
            continue

        access = data.get("access_token")
        if not access:
            continue
        expires_in = int(data.get("expires_in") or 3600)
        creds = XaiCredentials(
            auth_method="oauth",
            access_token=access,
            refresh_token=data.get("refresh_token"),
            expires_at=time.time() + expires_in,
            token_type=str(data.get("token_type") or "Bearer"),
            raw=data,
        )
        save_credentials(creds, home=home)
        with _poll_lock:
            if session_id in _poll_sessions:
                _poll_sessions[session_id]["status"] = "connected"
                _poll_sessions[session_id]["error"] = None
        logger.info("xAI OAuth login succeeded")
        return

    with _poll_lock:
        if session_id in _poll_sessions:
            _poll_sessions[session_id]["status"] = "error"
            _poll_sessions[session_id]["error"] = "expired_token"


def login_status(session_id: str | None = None, home: Path | None = None) -> dict[str, Any]:
    creds = load_credentials(home)
    out: dict[str, Any] = {"credentials": creds.to_public_dict()}
    if session_id:
        with _poll_lock:
            sess = _poll_sessions.get(session_id)
        if sess:
            out["session"] = {
                "session_id": session_id,
                "status": sess.get("status"),
                "error": sess.get("error"),
            }
            if sess.get("status") == "connected":
                out["credentials"] = load_credentials(home).to_public_dict()
        else:
            out["session"] = {
                "session_id": session_id,
                "status": "unknown",
                "error": None,
            }
    return out


def refresh_if_needed(home: Path | None = None) -> XaiCredentials:
    """Refresh OAuth access token if near expiry. Thread-safe."""
    with _refresh_lock:
        creds = load_credentials(home)
        if creds.auth_method != "oauth" or not creds.refresh_token:
            return creds
        if creds.expires_at and time.time() < float(creds.expires_at) - 120:
            return creds
        try:
            data = _http_form(
                TOKEN_URL,
                {
                    "grant_type": "refresh_token",
                    "refresh_token": creds.refresh_token,
                    "client_id": _client_id(),
                },
            )
        except RuntimeError as e:
            logger.warning("xAI token refresh failed: %s", e)
            return creds
        access = data.get("access_token")
        if not access:
            return creds
        expires_in = int(data.get("expires_in") or 3600)
        creds.access_token = access
        if data.get("refresh_token"):
            creds.refresh_token = data["refresh_token"]
        creds.expires_at = time.time() + expires_in
        save_credentials(creds, home=home)
        return creds


def resolve_bearer(home: Path | None = None) -> str | None:
    """Best available bearer for xAI API calls."""
    creds = refresh_if_needed(home)
    token = creds.bearer_token()
    if token:
        return token
    env_key = (
        os.environ.get("XAI_API_KEY", "").strip()
        or os.environ.get("REMEDY_XAI_API_KEY", "").strip()
    )
    return env_key or None

File: native/go/httpapi/xai_auth.go
package httpapi

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/AhmiDarrow/RemedyAI/native/go/secret"
)

const (
	xaiAccountsServer = "https://accounts.x.ai"
	xaiOAuthServer    = "https://auth.x.ai"
	xaiDeviceCodeURL  = xaiOAuthServer + "/oauth2/device/code"
	xaiTokenURL       = xaiOAuthServer + "/oauth2/token"
	xaiOAuthBuildID   = "auth.x.ai-device-v4-api-access"
	xaiDefaultScope   = "openid profile email offline_access grok-cli:access api:access"
	// Public xAI device-OAuth client id (override via REMEDY_XAI_OAUTH_CLIENT_ID).
	xaiDefaultClientID = "b1a00492-073a-47ea-816f-4c329264a828"
	xaiGrantDevice     = "urn:ietf:params:oauth:grant-type:device_code"
)

var xaiDeviceCodeCandidates = []string{
	xaiDeviceCodeURL,
	"https://auth.x.ai/oauth2/device/code",
}

type xaiCredentials struct {
	AuthMethod   string
	APIKey       string
	AccessToken  string
	RefreshToken string
	ExpiresAt    *float64
	TokenType    string
}

var (
	xaiPollMu       sync.Mutex
	xaiPollSessions = map[string]map[string]any{}
)

func xaiClientID() string {
	if v := strings.TrimSpace(os.Getenv("REMEDY_XAI_OAUTH_CLIENT_ID")); v != "" {
		return v
	}
	return xaiDefaultClientID
}

func xaiAuthPath(homeDir string) string {
	return filepath.Join(ResolveHomeDir(homeDir), "auth", "xai.json")
}

func coerceExpiresAt(v any) *float64 {
	if v == nil {
		return nil
	}
	switch t := v.(type) {
	case float64:
		return &t
	case int:
		f := float64(t)
		return &f
	case int64:
		f := float64(t)
		return &f
	case json.Number:
		f, err := t.Float64()
		if err != nil {
			return nil
		}
		return &f
	case string:
		s := strings.TrimSpace(t)
		if s == "" {
			return nil
		}
		var f float64
		if _, err := fmt.Sscanf(s, "%f", &f); err != nil {
			return nil
		}
		return &f
	default:
		return nil
	}
}

func (c xaiCredentials) connected() bool {
	switch c.AuthMethod {
	case "api_key":
		return strings.TrimSpace(c.APIKey) != ""
	case "oauth":
		if strings.TrimSpace(c.AccessToken) == "" {
			return false
		}
		if c.ExpiresAt != nil && time.Now().Unix() >= int64(*c.ExpiresAt)-60 {
			return strings.TrimSpace(c.RefreshToken) != ""
		}
		return true
	default:
		return false
	}
}

func (c xaiCredentials) toPublic(homeDir string) map[string]any {
	out := map[string]any{
		"provider":    "xai",
		"auth_method": c.AuthMethod,
		"connected":   c.connected(),
		"has_api_key": strings.TrimSpace(c.APIKey) != "",
		"has_oauth":   strings.TrimSpace(c.AccessToken) != "" || strings.TrimSpace(c.RefreshToken) != "",
		"expires_at":  nil,
	}
	if c.ExpiresAt != nil {
		out["expires_at"] = *c.ExpiresAt
	}
	enc := sealedAuthEncoding(xaiAuthPath(homeDir))
	out["credentials_encoding"] = enc
	if enc == "plain" && c.connected() {
		out["credentials_encoding_warning"] = "xAI credentials are stored as plaintext (DPAPI seal unavailable). " +
			"Anyone with access to your Windows user profile can read them. " +
			"Fix DPAPI / re-connect if this was unexpected."
	}
	return out
}

func loadXaiCredentials(homeDir string) xaiCredentials {
	path := xaiAuthPath(homeDir)
	data, err := readSealedJSON(path)
	if err != nil {
		envKey := strings.TrimSpace(os.Getenv("XAI_API_KEY"))
		if envKey == "" {
			envKey = strings.TrimSpace(os.Getenv("REMEDY_XAI_API_KEY"))
		}
		if envKey != "" {
			return xaiCredentials{AuthMethod: "api_key", APIKey: envKey, TokenType: "Bearer"}
		}
		return xaiCredentials{AuthMethod: "none", TokenType: "Bearer"}
	}
	method := strings.TrimSpace(fmt.Sprint(data["auth_method"]))
	if method == "" {
		method = "none"
	}
	return xaiCredentials{
		AuthMethod:   method,
		APIKey:       strings.TrimSpace(fmt.Sprint(nilToEmpty(data["api_key"]))),
		AccessToken:  strings.TrimSpace(fmt.Sprint(nilToEmpty(data["access_token"]))),
		RefreshToken: strings.TrimSpace(fmt.Sprint(nilToEmpty(data["refresh_token"]))),
		ExpiresAt:    coerceExpiresAt(data["expires_at"]),
		TokenType:    firstNonEmpty(strings.TrimSpace(fmt.Sprint(nilToEmpty(data["token_type"]))), "Bearer"),
	}
}

func nilToEmpty(v any) any {
	if v == nil {
		return ""
	}
	return v
}

func saveXaiCredentials(homeDir string, creds xaiCredentials) error {
	if _, err := authDir(homeDir); err != nil {
		return err
	}
	payload := map[string]any{
		"auth_method":   creds.AuthMethod,
		"api_key":       emptyToNil(creds.APIKey),
		"access_token":  emptyToNil(creds.AccessToken),
		"refresh_token": emptyToNil(creds.RefreshToken),
		"expires_at":    nil,
		"token_type":    firstNonEmpty(creds.TokenType, "Bearer"),
		"updated_at":    float64(time.Now().UnixNano()) / 1e9,
	}
	if creds.ExpiresAt != nil {
		payload["expires_at"] = *creds.ExpiresAt
	}
	return writeSealedJSON(xaiAuthPath(homeDir), payload)
}

func emptyToNil(s string) any {
	if strings.TrimSpace(s) == "" {
		return nil
	}
	return s
}

func clearXaiCredentials(homeDir string) {
	removeAuthFile(xaiAuthPath(homeDir))
}

func saveXaiAPIKey(homeDir, apiKey string) (xaiCredentials, error) {
	key := strings.TrimSpace(apiKey)
	if key == "" {
		return xaiCredentials{}, fmt.Errorf("API key is empty")
	}
	creds := xaiCredentials{AuthMethod: "api_key", APIKey: key, TokenType: "Bearer"}
	if err := saveXaiCredentials(homeDir, creds); err != nil {
		return xaiCredentials{}, err
	}
	return creds, nil
}

func xaiHTTPForm(endpoint string, form map[string]string) (map[string]any, error) {
	if strings.Contains(endpoint, "accounts.x.ai") &&
		(strings.Contains(endpoint, "/oauth2/device/code") || strings.Contains(endpoint, "/oauth2/token")) {
		return nil, fmt.Errorf(`{"error":"refusing accounts.x.ai for device/token API","url":%q,"hint":"Use https://auth.x.ai — accounts.x.ai returns 307 to /sign-in","oauth_build":%q}`,
			endpoint, xaiOAuthBuildID)
	}
	values := url.Values{}
	for k, v := range form {
		values.Set(k, v)
	}
	req, err := http.NewRequest(http.MethodPost, endpoint, strings.NewReader(values.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "RemedyDesktop-xAI-Auth/1.0 ("+xaiOAuthBuildID+")")
	client := &http.Client{
		Timeout: 30 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode >= 300 && resp.StatusCode < 400 {
		parsed := map[string]any{
			"error":       string(raw),
			"status":      resp.StatusCode,
			"request_url": endpoint,
			"oauth_build": xaiOAuthBuildID,
			"location":    resp.Header.Get("Location"),
			"hint": "xAI OAuth API redirected to a web sign-in page. " +
				"This usually means a stale local API binary is still calling " +
				"accounts.x.ai. Install a current build or replace remedy-runtime. " +
				"Expected host: auth.x.ai (build " + xaiOAuthBuildID + ").",
		}
		b, _ := json.Marshal(parsed)
		return nil, fmt.Errorf("%s", string(b))
	}
	var parsed map[string]any
	if len(raw) > 0 {
		if err := json.Unmarshal(raw, &parsed); err != nil {
			parsed = map[string]any{"error": string(raw), "status": resp.StatusCode}
		}
	}
	if parsed == nil {
		parsed = map[string]any{}
	}
	if resp.StatusCode >= 400 {
		parsed["status"] = resp.StatusCode
		parsed["request_url"] = endpoint
		parsed["oauth_build"] = xaiOAuthBuildID
		b, _ := json.Marshal(parsed)
		return nil, fmt.Errorf("%s", string(b))
	}
	return parsed, nil
}

func startXaiDeviceLogin(homeDir string) (map[string]any, error) {
	clientID := xaiClientID()
	scope := strings.TrimSpace(os.Getenv("REMEDY_XAI_OAUTH_SCOPE"))
	if scope == "" {
		scope = xaiDefaultScope
	}
	form := map[string]string{
		"client_id": clientID,
		"scope":     scope,
	}
	var (
		data    map[string]any
		lastErr error
		usedURL = xaiDeviceCodeURL
	)
	for _, u := range xaiDeviceCodeCandidates {
		got, err := xaiHTTPForm(u, form)
		if err != nil {
			lastErr = err
			continue
		}
		data = got
		usedURL = u
		break
	}
	if data == nil {
		if lastErr != nil {
			return nil, lastErr
		}
		return nil, fmt.Errorf(`{"error":"all device-code endpoints failed","oauth_build":%q}`, xaiOAuthBuildID)
	}
	deviceCode := strings.TrimSpace(fmt.Sprint(nilToEmpty(data["device_code"])))
	userCode := strings.TrimSpace(fmt.Sprint(nilToEmpty(data["user_code"])))
	if deviceCode == "" || userCode == "" {
		return nil, fmt.Errorf("Unexpected device-code response: %v", data)
	}
	interval := anyInt(data["interval"], 5)
	expiresIn := anyInt(data["expires_in"], 900)
	verificationURI := strings.TrimSpace(fmt.Sprint(nilToEmpty(data["verification_uri"])))
	if verificationURI == "" {
		verificationURI = strings.TrimSpace(fmt.Sprint(nilToEmpty(data["verification_url"])))
	}
	if verificationURI == "" {
		verificationURI = xaiAccountsServer + "/oauth2/device"
	}
	verificationComplete := strings.TrimSpace(fmt.Sprint(nilToEmpty(data["verification_uri_complete"])))
	if verificationComplete == "" {
		verificationComplete = verificationURI + "?user_code=" + url.QueryEscape(userCode)
	}
	sessionID := userCode
	home := ResolveHomeDir(homeDir)
	xaiPollMu.Lock()
	xaiPollSessions[sessionID] = map[string]any{
		"device_code": deviceCode,
		"client_id":   clientID,
		"interval":    interval,
		"expires_at":  float64(time.Now().Unix()) + float64(expiresIn),
		"home":        home,
		"status":      "pending",
		"error":       nil,
	}
	xaiPollMu.Unlock()
	go pollXaiUntilDone(sessionID)

	host := ""
	if u, err := url.Parse(usedURL); err == nil {
		host = u.Host
	}
	return map[string]any{
		"session_id":                 sessionID,
		"user_code":                  userCode,
		"device_code":                deviceCode,
		"verification_uri":           verificationURI,
		"verification_uri_complete":  verificationComplete,
		"interval":                   interval,
		"expires_in":                 expiresIn,
		"status":                     "pending",
		"oauth_host":                 host,
		"oauth_build":                xaiOAuthBuildID,
		"message": fmt.Sprintf(
			"Open %s and approve access, or go to %s and enter code %s.",
			verificationComplete, verificationURI, userCode,
		),
	}, nil
}

func pollXaiUntilDone(sessionID string) {
	xaiPollMu.Lock()
	sess := cloneAnyMap(xaiPollSessions[sessionID])
	xaiPollMu.Unlock()
	if len(sess) == 0 {
		return
	}
	deviceCode, _ := sess["device_code"].(string)
	clientID, _ := sess["client_id"].(string)
	interval := anyInt(sess["interval"], 5)
	if interval < 3 {
		interval = 3
	}
	expiresAt := anyFloat(sess["expires_at"], float64(time.Now().Unix())+900)
	home, _ := sess["home"].(string)

	for float64(time.Now().Unix()) < expiresAt {
		time.Sleep(time.Duration(interval) * time.Second)
		data, err := xaiHTTPForm(xaiTokenURL, map[string]string{
			"grant_type":  xaiGrantDevice,
			"device_code": deviceCode,
			"client_id":   clientID,
		})
		if err != nil {
			errCode := err.Error()
			if strings.Contains(errCode, "authorization_pending") || strings.Contains(errCode, "slow_down") {
				if strings.Contains(errCode, "slow_down") {
					interval = min(interval+2, 15)
				}
				continue
			}
			if strings.Contains(errCode, "expired") || strings.Contains(errCode, "access_denied") {
				xaiPollMu.Lock()
				if row, ok := xaiPollSessions[sessionID]; ok {
					row["status"] = "error"
					row["error"] = errCode
				}
				xaiPollMu.Unlock()
				return
			}
			continue
		}
		access := strings.TrimSpace(fmt.Sprint(nilToEmpty(data["access_token"])))
		if access == "" {
			continue
		}
		expiresIn := anyInt(data["expires_in"], 3600)
		exp := float64(time.Now().Unix()) + float64(expiresIn)
		creds := xaiCredentials{
			AuthMethod:   "oauth",
			AccessToken:  access,
			RefreshToken: strings.TrimSpace(fmt.Sprint(nilToEmpty(data["refresh_token"]))),
			ExpiresAt:    &exp,
			TokenType:    firstNonEmpty(strings.TrimSpace(fmt.Sprint(nilToEmpty(data["token_type"]))), "Bearer"),
		}
		_ = saveXaiCredentials(home, creds)
		xaiPollMu.Lock()
		if row, ok := xaiPollSessions[sessionID]; ok {
			row["status"] = "connected"
			row["error"] = nil
		}
		xaiPollMu.Unlock()
		return
	}
	xaiPollMu.Lock()
	if row, ok := xaiPollSessions[sessionID]; ok {
		row["status"] = "error"
		row["error"] = "expired_token"
	}
	xaiPollMu.Unlock()
}

func xaiLoginStatus(homeDir, sessionID string) map[string]any {
	creds := loadXaiCredentials(homeDir)
	out := map[string]any{
		"credentials": creds.toPublic(homeDir),
	}
	sessionID = strings.TrimSpace(sessionID)
	if sessionID == "" {
		return out
	}
	xaiPollMu.Lock()
	sess := xaiPollSessions[sessionID]
	xaiPollMu.Unlock()
	if sess == nil {
		out["session"] = map[string]any{
			"session_id": sessionID,
			"status":     "unknown",
			"error":      nil,
		}
		return out
	}
	status, _ := sess["status"].(string)
	out["session"] = map[string]any{
		"session_id": sessionID,
		"status":     status,
		"error":      sess["error"],
	}
	if status == "connected" {
		out["credentials"] = loadXaiCredentials(homeDir).toPublic(homeDir)
	}
	return out
}

func cloneAnyMap(in map[string]any) map[string]any {
	if in == nil {
		return nil
	}
	out := make(map[string]any, len(in))
	for k, v := range in {
		out[k] = v
	}
	return out
}

func anyFloat(v any, def float64) float64 {
	switch t := v.(type) {
	case float64:
		return t
	case int:
		return float64(t)
	case int64:
		return float64(t)
	default:
		return def
	}
}

func (s *Server) handleXaiAuthStatus(w http.ResponseWriter, _ *http.Request) {
	home := ResolveHomeDir(s.homeDir)
	writeJSON(w, http.StatusOK, loadXaiCredentials(home).toPublic(home))
}

func (s *Server) handleXaiOAuthMeta(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, map[string]any{
		"oauth_build":     xaiOAuthBuildID,
		"device_code_url": xaiDeviceCodeURL,
		"token_url":       xaiTokenURL,
		"accounts_server": xaiAccountsServer,
	})
}

func (s *Server) handleXaiAuthLogin(w http.ResponseWriter, _ *http.Request) {
	home := ResolveHomeDir(s.homeDir)
	result, err := startXaiDeviceLogin(home)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, map[string]string{
			"detail": "xAI OAuth start failed: " + err.Error(),
		})
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (s *Server) handleXaiAuthLoginStatus(w http.ResponseWriter, r *http.Request) {
	home := ResolveHomeDir(s.homeDir)
	sessionID := strings.TrimSpace(r.URL.Query().Get("session_id"))
	// GET must not steal the chat provider (Python only refreshes when already on xAI).
	writeJSON(w, http.StatusOK, xaiLoginStatus(home, sessionID))
}

func (s *Server) handleXaiAuthAPIKey(w http.ResponseWriter, r *http.Request) {
	var req struct {
		APIKey string `json:"api_key"`
	}
	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"detail": "invalid JSON body"})
		return
	}
	home := ResolveHomeDir(s.homeDir)
	creds, err := saveXaiAPIKey(home, req.APIKey)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"detail": err.Error()})
		return
	}
	key := strings.TrimSpace(req.APIKey)
	cfg := LoadConfig(s.homeDir)
	prev := strings.ToLower(strings.TrimSpace(cfgString(cfg, "llm_provider", "")))
	if prev == "xai" || prev == "" {
		provider, model, baseURL := normalizeLLMSettings("xai", cfgString(cfg, "llm_model", ""), cfgString(cfg, "llm_base_url", "https://api.x.ai/v1"))
		path := FindConfigPath(s.homeDir)
		if path == "" {
			path = DefaultConfigPath(s.homeDir)
		}
		if path != "" {
			if cfg == nil {
				cfg = ConfigMap{}
			}
			cfg["llm_provider"] = provider
			cfg["llm_model"] = model
			cfg["llm_base_url"] = baseURL
			cfg["llm_api_key"] = ""
			_ = WriteConfig(path, scrubConfigSecrets(cfg))
		}
		_ = secret.SetProviderSecret(home, "xai", key)
	}
	out := creds.toPublic(home)
	out["status"] = "saved"
	writeJSON(w, http.StatusOK, out)
}

func (s *Server) handleXaiAuthLogout(w http.ResponseWriter, _ *http.Request) {
	home := ResolveHomeDir(s.homeDir)
	clearXaiCredentials(home)
	cfg := LoadConfig(s.homeDir)
	if strings.EqualFold(cfgString(cfg, "llm_provider", ""), "xai") {
		path := FindConfigPath(s.homeDir)
		if path != "" {
			cfg["llm_api_key"] = ""
			_ = WriteConfig(path, scrubConfigSecrets(cfg))
		}
		_ = secret.SetProviderSecret(home, "xai", "")
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"status":    "logged_out",
		"provider":  "xai",
		"connected": false,
	})
}
